src/pyunicorn/funcnet/_ext/numerics.pyx

```diff
@@ -98,34 +94,129 @@ def _cross_correlation_all(
+        ndarray[DFIELD_t, ndim=1, mode='c'] dxyzarray = np.zeros(k+1, dtype=DFIELD)
+        ndarray[NODE_t, ndim=1, mode='c'] k_xz = np.zeros(T, dtype=NODE)
+        ndarray[NODE_t, ndim=1, mode='c'] k_yz = np.zeros(T, dtype=NODE)
+        ndarray[NODE_t, ndim=1, mode='c'] k_z = np.zeros(T, dtype=NODE)
+
+    # Loop over time
+    for i in range(T):
+        # Growing cube algorithm:
+        # Test if n = #(points in epsilon-environment of reference point i) > k
+        # Start with epsilon for which 95% of points are inside the cube
+        # for a multivariate Gaussian
+        # eps increased by 2 later, also the initial eps
+        eps = (k/T)**(1./dim)
+
+        # n counts the number of neighbors
+        n = 0
+        while n <= k:
+            # Increase cube size
+            eps *= 2.
+            # Start with zero again
+            n = 0
+            # Loop through all points
+            for t in range(T):
+                d = 0
+                while d < dim and abs(array[d,i] - array[d,t]) < eps:
+                    d += 1
+
+                # If all distances are within eps, the point t lies
+                # within eps and n is incremented
+                if d == dim:
+                    indexfound[n] = t
+                    n += 1
+
+        # Calculate distance to points only within epsilon environment
+        # according to maximum metric
+        for j in range(n):
+            index = indexfound[j]
+
+            # calculate maximum metric distance to point
+            dxyz = 0.
+            for d in range(dim):
+                dist[d,j] = abs(array[d,i] - array[d,index])
+                dxyz = max(dist[d,j], dxyz)
+
+            # insertion-sort current distance into 'dxyzarray'
+            # if it is among the currently smallest k+1 distances
+            if j == 0:
+                dxyzarray[j] = dxyz
+            else:
+                m = min(k, <int> (j-1))
+                # go through previously sorted smallest distances and
+                # if it is smaller than any, find slot for current distance
+                while m >= 0 and dxyz < dxyzarray[m]:
+                    # if it's not in the last slot already,
+                    # move previously found distance to the right
+                    if not m == k:
+                        dxyzarray[m+1] = dxyzarray[m]
+                    m -= 1
+
+                # sort in, if a slot was found
+                if not m == k:
+                    dxyzarray[m+1] = dxyz
+
+        # Epsilon of k-th nearest neighbor in joint space
+        epsmax = dxyzarray[k]
+
+        # Count neighbors within epsmax in subspaces, since the reference
+        # point is included, all neighbors are at least 1
+        kz = 0
+        kxz = 0
+        kyz = 0
+        for j in range(T):
+
+            # X-subspace
+            dx = abs(array[0,i] - array[0,j])
+            for d in range(1, dim_x):
```

Comment:
within `CouplingAnalyis`, `dim_x` (and `dim_y`) will always be `1` (see [here](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46L345)/[here](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46L666) and [here](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46L666)), hence this loop is never executed, and the respective loops for dimensions `y` and `z` are only executed once

Reply:
similarly, in the algorithm in `CouplingAnalysis.mutual_information()` is adapted from `CouplingAnalysis.information_transfer()`, but containing some unecessary twists: 
- [`max_lag`](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46R337) will always be equal to `tau_max` there, 
- [`dim`](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46R338) will always be `2`, and 
- the subsequent [`loop`](https://github.com/pik-copan/pyunicorn/pull/195/files#diff-e83ca16d7669e9955f8a9a51836b17d72aac6f0a1372f818b7163a7e3bacfe46R341) could therefore be replaced by just two lines of code.

Reply:
Ah. I should have fully read your latest PR comment first.